### PR TITLE
27 apis for vendors

### PR DESCRIPTION
**Pull request by Qmarapatla**
All vendor models are present, please review the code.

### CODE REVIEW

File: apps/vendor/models.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    code = models.CharField(max_length=50, null=True, default=None)
+    name = models.CharField(max_length=255, null=True, default=None)
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+    def __str__(self):
+        return f'{self.vendor_category_id}_{self.name} ({self.code})'
+
+    class Meta:
+        db_table = 'vendor_category'
```
Comment (thulasiram1216) [inline apps/vendor/models.py:20]: db_table value should come from utils_variables.py

File: apps/vendor/models.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    payment_cycle = models.CharField(max_length=255, null=True, default=None)
+    run_on = models.CharField(max_length=255, null=True, default=None)
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+    def __str__(self):
+        return f'{self.payment_term_id}_{self.name} ({self.code})'
+    
+    class Meta:
+        db_table = 'vendor_payment_terms'
```
Comment (thulasiram1216) [inline apps/vendor/models.py:38]: db_table value should come from utils_variables.py

File: apps/vendor/models.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+class VendorPaymentTerms(models.Model):
+    payment_term_id = models.AutoField(primary_key=True)
+    name = models.CharField(max_length=255, null=True, default=None)
+    code = models.CharField(max_length=50, null=True, default=None)
+    fixed_days = models.IntegerField(null=True, default=None)
+    no_of_fixed_days = models.IntegerField(null=True, default=None)
+    payment_cycle = models.CharField(max_length=255, null=True, default=None)
+    run_on = models.CharField(max_length=255, null=True, default=None)
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
```
Comment (thulasiram1216) [inline apps/vendor/models.py:32]: updated_at should satisfy below statement, did you test?
updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP

File: apps/vendor/models.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+    def __str__(self):
+        return f'{self.vendor_agent_id}_{self.name} ({self.code})'
+    
+    class Meta:
+        db_table = 'vendor_agent'
+
+
+def custom_upload_to(instance, filename):
+    file_extension = filename.split('.')[-1]
```
Comment (thulasiram1216) [inline apps/vendor/models.py:69]: Why this function is outside the class? if it is common function then move it to utils_methods.py

File: apps/vendor/models.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    bank_name = models.CharField(max_length=255, null=True, default=None)
+    branch = models.CharField(max_length=255, null=True, default=None)
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+    
+    def __str__(self):
+        return f'{self.vendor_id}_{self.name} ({self.code})'
+
+    class Meta:
+        db_table = 'vendor'
```
Comment (thulasiram1216) [inline apps/vendor/models.py:126]: db_table value should come from utils_variables.py

File: apps/vendor/models.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+class VendorAttachment(models.Model):
+    attachment_id = models.AutoField(primary_key=True)
+    vendor_id = models.ForeignKey(Vendor, on_delete=models.CASCADE, null=True, default=None, db_column='vendor_id')
+    attachment_name = models.CharField(max_length=255, null=True, default=None)
+    attachment_path = models.CharField(max_length=255, null=True, default=None)
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+    class Meta:
+        db_table = 'vendor_attachments'  # Set custom table name
```
Comment (thulasiram1216) [inline apps/vendor/models.py:138]: db_table value should come from utils_variables.py
/

File: apps/vendor/models.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    phone = models.CharField(max_length=50, null=True, default=None)
+    email = models.EmailField(max_length=255, null=True, default=None)
+    longitude = models.DecimalField(max_digits=10, decimal_places=6, null=True, default=None)
+    latitude = models.DecimalField(max_digits=10, decimal_places=6, null=True, default=None)
+    route_map = models.CharField(max_length=255, null=True, default=None)
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+    class Meta:
+        db_table = 'vendor_addresses'
```
Comment (thulasiram1216) [inline apps/vendor/models.py:165]: db_table value should come from utils_variables.py

File: apps/vendor/urls.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+#add your urls 
+
+router = routers.DefaultRouter()
+router.register(r'vendor', VendorsView)
+router.register(r'vendor_category', VendorCategoryView)
+router.register(r'vendor_payment_terms', VendorPaymentTermsView)
+router.register(r'vendor_agent', VendorAgentView)
+router.register(r'vendor_attachment', VendorAttachmentView)
+router.register(r'vendor_address', VendorAddressView)
```
Comment (thulasiram1216) [inline apps/vendor/urls.py:13]: Please follow the same pattern where pavan and pramod followed.

File: apps/vendor/models.py (hunk 2 above)
Comment (Qmarapatla) [reply apps/vendor/models.py:38]: sure

File: apps/vendor/models.py (hunk 4 above)
Comment (Qmarapatla) [reply apps/vendor/models.py:69]: I move Thulasi.

File: apps/vendor/models.py (hunk 5 above)
Comment (Qmarapatla) [reply apps/vendor/models.py:126]: ok

File: apps/vendor/models.py (hunk 6 above)
Comment (Qmarapatla) [reply apps/vendor/models.py:138]: ok

File: apps/vendor/models.py (hunk 7 above)
Comment (Qmarapatla) [reply apps/vendor/models.py:165]: ok

File: apps/vendor/urls.py (hunk 8 above)
Comment (Qmarapatla) [reply apps/vendor/urls.py:13]: ok

File: apps/vendor/models.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+class VendorAgent(models.Model):
+    vendor_agent_id = models.AutoField(primary_key=True)
+    code = models.CharField(max_length=50, null=True, default=None)
+    name = models.CharField(max_length=255, null=True, default=None)
+    commission_rate = models.DecimalField(max_digits=18, decimal_places=2, null=True, default=None)
+    
+    RATE_ON_CHOICES = [
+        ('Qty', 'Quantity'),
+        ('Amount', 'Amount'),
```
Comment (thulasiram1216) [inline apps/vendor/models.py:49]: This should have another value 'Both' Can you double check

File: apps/vendor/models.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    linked_in = models.URLField(max_length=255, null=True, default=None)
+    payment_term_id = models.ForeignKey(VendorPaymentTerms,on_delete=models.CASCADE, null=True, default=None, db_column='payment_term_id')
+    price_category_id = models.ForeignKey(PriceCategories, on_delete=models.CASCADE, null=True, default=None, db_column='price_category_id')
+    vendor_agent_id = models.ForeignKey(VendorAgent, on_delete=models.CASCADE, null=True, default=None, db_column='vendor_agent_id')
+    transporter_id = models.ForeignKey(Transporters, on_delete=models.CASCADE, null=True, default=None, db_column='transporter_id')
+    credit_limit = models.DecimalField(max_digits=18, decimal_places=2, null=True, default=None)
+    max_credit_days = models.IntegerField(null=True, default=None)
+    interest_rate_yearly = models.DecimalField(max_digits=18, decimal_places=2, null=True, default=None)
+    rtgs_ifsc_code = models.CharField(max_length=255, null=True, default=None)
+    accounts_number = models.CharField(max_length=255, null=True, default=None)
```
Comment (thulasiram1216) [inline apps/vendor/models.py:109]: Can you check with pavan and apply the encrypt/decrypt that he applied to account number field, Also change variable name to account_number

File: apps/vendor/models.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+    ADDRESS_TYPE_CHOICES = [
+        ('billing', 'Billing'),
+        ('shipping', 'Shipping')
+        ]
+    address_type = models.CharField(max_length=10, choices=ADDRESS_TYPE_CHOICES, null=True, default=None)
+    address = models.CharField(max_length=255, null=True, default=None)
+    country = models.CharField(max_length=255, null=True, default=None)
+    state = models.CharField(max_length=255, null=True, default=None)
+    city = models.CharField(max_length=255, null=True, default=None)
+    pin_code = models.CharField(max_length=50, null=True, default=None)
```
Comment (thulasiram1216) [inline apps/vendor/models.py:148]: Above state, city, country fields are modified. you need to pull the latest changes from main and change your code accordingly

File: apps/vendor/models.py (hunk 9 above)
Comment (Qmarapatla) [reply apps/vendor/models.py:49]: Following is the SQL structure Thulasi, there is no 'both' option.
    rate_on ENUM("Qty", "Amount"),
    amount_type ENUM("Taxable", "BillAmount"),

File: apps/vendor/models.py (hunk 11 above)
Comment (Qmarapatla) [reply apps/vendor/models.py:148]: updated Thulasi

File: apps/vendor/models.py (hunk 10 above)
Comment (Qmarapatla) [reply apps/vendor/models.py:109]: updated Thulasi

**Pull request merged**